--- a/drivers/clk/mediatek/clk-mt8163-mmsys.c
+++ b/drivers/clk/mediatek/clk-mt8163-mmsys.c
@@ -94,11 +94,18 @@
 };
 MODULE_DEVICE_TABLE(platform, clk_mt8163_mm_id_table);
 
+static const struct of_device_id of_match_clk_mt8163_mm[] = {
+	{ .compatible = "mediatek,mt8163-mmsys", .data = &mm_desc },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, of_match_clk_mt8163_mm);
+
 static struct platform_driver clk_mt8163_mm_drv = {
 	.probe = mtk_clk_pdev_probe,
 	.remove = mtk_clk_pdev_remove,
 	.driver = {
 		.name = "clk-mt8163-mmsys",
+		.of_match_table = of_match_clk_mt8163_mm,
 	},
 	.id_table = clk_mt8163_mm_id_table,
 };
